### DISCUSSION TITLE
MSW is not intercepting calls. How to debug?

### DISCUSSION THREAD

**Discussion by michalHT** [mswjs/msw#576] [Q&A] [answered]
What could be the reason for the following behavior? Is there a way to see what's actually failing?
- "[MSW] Mocking enabled." is logged
- `worker.printHandlers();` loggs out "[rest] POST https://myapi.localtest.me/api/orders"
- `POST https://myapi.localtest.me/api/orders` call inside the app is not intercepted by the worker
- [Path matching REPL](https://mswjs.io/repl/?path=https%3A%2F%2Fmyapi.localtest.me%2Fapi%2Forders&url=https%3A%2F%2Fmyapi.localtest.me%2Fapi%2Forders%2F) does not find any problems
- `onUnhandledRequest` does not log anything regardless of whether "error" "warn" or custom callback is passed
- I can see the worker running in browser dev tools
- The same handlers are used for setting up mocking via `setupServer` for testing in Jest and it works great there
- (https://myapi.localtest.me is served by local IIS server)

**Comment by timdeschryver** [comment, marked as answer]
Hi @michalHT , can you retry this with the latest version (0.26.1), this version [includes a suggestion helper](https://github.com/mswjs/msw/pull/560) that might help to debug this issue. If that doesn't help, could you create a reproduction so we can take a look?

[Release notes](https://github.com/mswjs/msw/releases/tag/v0.26.1)

**Reply by michalHT** [reply]
I installed 0.26.1 but I can't see any changes. Any particular option I need to turn on to get these suggestions?

Here's what I have:
```javascript
await worker.start({
  onUnhandledRequest: "error",
  serviceWorker: {
    url: "/Static/mockServiceWorker.js",
  },
});

worker.printHandlers();
```

With this setup handlers are printed out correctly, but the calls are not intercepted and there are no errors in the console.
I don't think I'll be able to create a reproduction because the same code works well on CRA. And here I'm working on a custom webpack config with IIS as a server. 

So the main problem is that I don't know what is actually failing since there are no errors from MSW. Is there any webpack or server configuration that could make MSW fail?

**Reply by kettanaito** [reply]
Beware that since you're registering the worker at `/Static/` it can only intercept the requests to the child routes of `/Static/*`.  For example, these requests will be intercepted by the worker:

- `/Static/user`
- `/Static/payment/abc-123`

While these _will not_:

- `/user`
- `/payment/abc-123`

I trust that is the reason requests are not intercepted and there are no errors: they never reach the worker as they are outside of the worker's scope. Here's a great excerpt from the [Using Service Workers (MDN)](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API/Using_Service_Workers):

> Also note:
> 
> - The service worker will only catch requests from clients under the service worker's scope.
> - The max scope for a service worker is the location of the worker.
> - If your service worker is active on a client being served with the Service-Worker-Allowed header, you can specify a list of max scopes for that worker.
> - In Firefox, Service Worker APIs are hidden and cannot be used when the user is in private browsing mode.

You can see the scope that the Service Worker is registered at in your browser's DevTools. For example, in Chrome that would be here: chrome://serviceworker-internals/?devtools

## How do I resolve this?

### Option 1. Register the worker at the root level of your app.

(_Recommended_) In other words, so that the worker script is available at `/mockServiceWorker.js` (notice no preceding path after `/`). This will make the worker's scope to control all requests on the page, which is most likely what you want. 

### Option 2. Use the `Service-Worker-Allowed` header.

If you're in charge of your app's dev server, you can set the `Service-Worker-Allowed` header on each response of your server. By providing that header with a root of your server (`/`) you will let the browser know that even the workers that aren't normally eligible to handle a request out of their scope can still do that. 

Here's how we append the `Service-Worker-Allowed` header in our Webpack Dev Server used for testing MSW:

https://github.com/mswjs/msw/blob/9d0d78c9f10fe344f7a86f12325a0ff20a4c8ab4/test/support/spawnServer.ts#L108-L112

> Appending a fixed header to all requests can also be done in any other server framework you may use. 

With this option, you can keep your worker script in `/Static`, but you'd have to meddle with the dev server setup, which isn't the best solution in case you are not familiar with it.

**Reply by michalHT** [reply]
Thank you very much for this elaborate response. Much appreciated!
It made me realize there are 2 more important details I forgot to mention:
- App URL is different than API url (https://**app**.localtest.me/ vs https://**myapi**.localtest.me/).
- I'm setting https://**myapi**.localtest.me/` as a base url for handlers using [this approach](https://github.com/mswjs/msw/pull/406#issuecomment-708519619)

Is MSW able to handle such scenario or does it mean that even if I place the worker file under https://**app**.localtest.me/mockServiceWorker.js it won't be able to intercept calls to https://**myapi**.localtest.me/?

And if the latter is true, is there a suggested workaround?

**Reply by kettanaito** [reply]
As long as the request is performed from the origin that the worker can control (from `https://app.localtest.me`), MSW will intercept that request. It's not the destination of the request that matters, but the origin. 

Have you tried moving the worker script higher in the scope? That should eliminate the issue, from what I can see. I doubt your request was happening from the `/Static/` page. 

Let me add some clarifications of what the "scope" actually means:

Let's say you register your worker at `/product` page (`url: "/product/mockServiceWorker.js"`). This means that _whichever_ request occurs when you are on `/product` or its child pages (i.e. `/product/abc-123`) will be intercepted and handled. However, if you move to a homepage `/` or a different page `/about`, none of the requests that happen there will be caught. Those pages lie outside of the worker's scope (`/product`). 

I think the only scenario when a request isn't intercepted and there are no errors when using `onUnhandledRequest: "error"` is when the worker's scope cannot affect the origin of that request.

**Reply by michalHT** [reply]
All right, that clarifies a lot! This is some stellar support for open source library. Hats off to you!

I will try to put the worker file in the root of the app. This is not as simple as one would think in this app, but I'll do my best and post and update soon.

**Reply by michalHT** [reply]
It worked like a charm after placing worker file in the app root. 
Thank you for all the help Artem. You are awesome!

**Reply by kettanaito** [reply]
Glad it helped, @michalHT. Please, mark the comment that you think helped you resolve the issue with the Answer icon (checkmark) next to it. Thanks :)

If you'd like to support me in answering yours and other people's questions, consider sponsoring the project by clicking on the "[Sponsor](https://github.com/sponsors/kettanaito)" button at the top of the repo!

**Comment by kettanaito** [comment]
To help our users resolve issues like this, we've published a [Debugging uncaught requests](https://mswjs.io/docs/recipes/debugging-uncaught-requests) recipe in our documentation!

**Comment by mod7ex** [comment]
i'm using vite and vitest and almost msw isn't recognized, requests are not intercepted

**Reply by nilabhanand** [reply]
Were you able to figure it out? I'm also getting the issue while using vite & vitest. The requests are not being intercepted despite being defined in the handlers.

**Reply by nilabhanand** [reply]
MSW intercepts the requests in the browser but during testing using vitest, the requests don't get intercepted.

**Reply by kettanaito** [reply]
Hey. Could somebody please post a link to a reproduction repository? Thanks.

**Reply by nilabhanand** [reply]
Hey I was able to figure it out. Had to add `fetch` from cross-fetch into polyfill

**Reply by ethanppl** [reply]
For others: 
```typescript
import { afterAll, afterEach, beforeAll } from 'vitest';
import { setupServer } from 'msw/node';
import { fetch } from 'cross-fetch';

const server = setupServer([/* handlers */]);

// Need to add this line
global.fetch = fetch;

beforeAll(() => server.listen({ onUnhandledRequest: 'error' }));
afterAll(() => server.close());
afterEach(() => server.resetHandlers());
```

**Reply by charlieforward9** [reply]
I am having an issue where the first MSW call works and the next one does not, failing. 

If I remove the first network call, the second one works...

I am using Vitest, Amplify and Graphql Codegen to make my msw handlers and mock data.

**Reply by cilladev** [reply]
> I am having an issue where the first MSW call works and the next one does not, failing.
> 
> If I remove the first network call, the second one works...
> 
> I am using Vitest, Amplify and Graphql Codegen to make my msw handlers and mock data.

Currently having this same issue - where you able to find a solution for this ?
